genesis/recorders/file_writers.py

```diff
@@ -0,0 +1,351 @@
+            if self.fourcc not in ["avc1", "mp4v", "h264"]:
+                gs.raise_exception(
+                    f"[{type(self).__name__}] '{self.fourcc}' is not a supported video codec for .mp4, "
+                    "expected 'avc1', 'mp4v', or 'h264'"
+                )
+        elif self.filename.endswith(".avi"):
+            if self.fourcc not in ["XVID", "MJPG"]:
+                gs.raise_exception(
+                    f"[{type(self).__name__}] '{self.fourcc}' is not a supported video codec for .avi, "
+                    "expected 'XVID' or 'MJPG'"
+                )
+        else:
+            gs.raise_exception(f"[{type(self).__name__}] Video filename must end with .mp4 or .avi.")
+
+
+@register_recording(Cv2VideoFileWriterOptions)
+class Cv2VideoFileWriter(BaseVideoFileWriter):
+
+    def initialize(self):
+        super().initialize()
+        self.video_writer: cv2.VideoWriter | None = None
+
+        # lazy init these values when processing the first frame
+        self.width, self.height = 0, 0
+        self.is_color: bool | None = None
+
+        os.makedirs(os.path.abspath(os.path.dirname(self._options.filename)), exist_ok=True)
+
+    def _initialize_data(self, data):
+        self.is_color = len(data[0].shape) == 3 and data[0].shape[-1] == 3
+        if not self.is_color and len(data[0].shape) != 2:
+            gs.raise_exception(f"[{type(self).__name__}] Data must be either grayscale [H, W] or color [H, W, RGB]")
+        self.height, self.width = data[0].shape[:2]
+
+    def _initialize_writer(self):
+        filename = self._get_filename()
+        video_writer = cv2.VideoWriter(
+            filename,
+            cv2.VideoWriter_fourcc(*self._options.fourcc),
+            self.fps,
+            (self.width, self.height),
+            self.is_color,
+        )
+        if not video_writer.isOpened():
+            gs.raise_exception(f'[{type(self).__name__}] Failed to open video writer for "{filename}"')
+
+    def process(self, data, cur_time):
+        if self.is_color is None:
+            self._initialize_data(data)
+        if self.video_writer is None:
+            self._initialize_writer()
+
+        if self.is_color:
+            data = cv2.cvtColor(data, cv2.COLOR_RGB2BGR)
+        self.video_writer.write(data)
+
+    def cleanup(self):
+        if self.video_writer:
+            self.video_writer.release()
+            gs.logger.info(f'Video saved to "~<{self._options.filename}>~".')
+            self.video_writer = None
+
+    @property
+    def run_in_thread(self) -> bool:
+        return True
+
+
+class CSVFileWriterOptions(BaseFileWriterOptions):
+    """
+    Writes to a .csv file using `csv.writer`.
+
+    Can handle any array-like or dict[str, array-like] output, e.g. from sensors.
+
+    Parameters
+    ----------
+    filename : str
+        The name of the CSV file to save the data.
+    header : list[str] | None, optional
+        Column headers for the CSV file.
```

Comment:
This is clearly not sufficient. You should clearly explain how data gets formatted in a row:
* If passing a dict, it cannot be further nested and values are processed in order
* Values must be N-dimensional tensors, arrays or scalars (np.generic, int, float, str)
* All tensors and arrays are flattened in C-order (aka row-major)

Each resulting scalar element must have its own name in header.